distribution/push_v2.go

```diff
@@ -360,6 +364,18 @@ func (pd *v2PushDescriptor) Upload(ctx context.Context, progressOutput progress.
 				return distribution.Descriptor{}, xfer.DoNotRetry{Err: err}
 			}
 			return err.Descriptor, nil
+		case errcode.Errors:
+			for _, e := range err {
+				switch e := e.(type) {
+				case errcode.Error:
+					if e.Code == errcode.ErrorCodeUnauthorized && !pd.pushState.hasAuthInfo {
+						// continue to try other candidate as other repos may not require credentials to mount the blob
+						logrus.Debugf("failed to push layer to registry because unauthorized error and user not login")
+						continue candidateLoop
```

Comment:
Does the layer upload need to be cancelled here? Maybe instead of a continue statement just ensure that the remove doesn't happen so there are fewer early exit/continue conditions.

Reply:
![image](https://user-images.githubusercontent.com/1619030/37375346-178213b0-275a-11e8-8949-887f149be837.png)

this is my first attempt, what do you think of this?

Reply:
as @aaronlehmann mentioned 
> I think this should continue instead of break. There might be other repos which do not require credentials to mount the blob. The continue can go inside the error checking code, instead of setting an isUnauthorizedError variable (it would mean having a label on the candidate loop and using continue candidateLoop).

is that the reason when authorizedError happened, it should continue to check other candidates?

Reply:
the main difference is do we break here when first authorizedError met or continue to try other candidates

Reply:
do we have some conclusion on this so that I can have another commit before this weekend?  @dmcgowan @aaronlehmann  : )

Reply:
@dmcgowan is right that this needs to cancel the layer upload.

How about adding `&& !isUnauthorizedError` to `if len(mountCandidate.SourceRepository) > 0...` below, instead of the continue?

Reply:
> How about adding && !isUnauthorizedError to if len(mountCandidate.SourceRepository) > 0... below, instead of the continue?

it looks like it will still continue to try next candidate which is not necessary?